### ISSUE TITLE
Signed distance behaviour for enclosed polygons

### ISSUE THREAD

**Issue by hsaunders1904** [Fusion-Power-Plant-Framework/bluemira#2273]
## Describe the bug

When the subject polygon passed into `signed_distance_2D_polygon` is completely within the target polygon, the signed distance is positive.

The function has the following signature and descriptions:

```python
def signed_distance_2D_polygon(subject_poly: np.ndarray, target_poly: np.ndarray) -> np.ndarray:
    """
    2-D vector-valued signed distance function from a subject polygon to a target
    polygon. The return values are negative for points outside the subject polygon, and
    positive for points inside the subject polygon.
    """
```

If the subject polygon is inside the target polygon, I would expect this to mean the signed distance would be completely negative, as every point in the target polygon would be outside the subject. The actual behaviour is that the signed distance is completely positive.

The converse is also true, but unexpected. If we take the same scenario, but swap the subject and target polygons, the signed distance is completely negative. We'd expect this to be positive, as every point in the target polygon is within the subject.

## Steps to reproduce

```pycon
>>> from bluemira.display import plot_2d
>>> from bluemira.geometry.tools import signed_distance_2D_polygon, make_circle

>>> small_circle = make_circle(5, center=(0, 0, 0), axis=(0, 1, 0))
>>> big_circle = make_circle(10, center=(0, 0, 0), axis=(0, 1, 0))
>>> plot_2d([small_circle, big_circle])
```

<img src=https://github.com/Fusion-Power-Plant-Framework/bluemira/assets/33317174/0bee6dcb-2e82-4596-b96d-ac355cee63d3 width=500>

```pycon
>>> signed_distance_2D_polygon(
...     small_circle.discretize(20).xz.T, big_circle.discretize(20).xz.T
... ).T
[4.93180652 4.93180652 4.93180652 4.93180652 4.93180652 4.93180652
 4.93180652 4.93180652 4.93180652 4.93180652 4.93180652 4.93180652
 4.93180652 4.93180652 4.93180652 4.93180652 4.93180652 4.93180652
 4.93180652 4.93180652]
```

Now swapping the subject and target:

```pycon
>>> signed_distance_2D_polygon(
...     big_circle.discretize(20).xz.T, small_circle.discretize(20).xz.T
... ).T
[-5. -5. -5. -5. -5. -5. -5. -5. -5. -5. -5. -5. -5. -5. -5. -5. -5. -5.
 -5. -5.]
```

## Expected behaviour

I was expecting that if the subject polygon is completely enclosed by the target, then every point of the target polygon is outside the subject, and we would get a negative signed distance. Similar for the converse.

## Additional context

Let me know if this makes sense, this has been driving me a bit crazy trying to get the keep-in/out zones working for the new optimiser interface. Entirely possible that I've made a mistake somewhere, I've honestly started to lose track 😅!

**Comment by CoronelBuendia** [comment]
Hmm... this is a stream of conscience reply, apologies in advance:

As a reminder of what the SDF is:
```python
from bluemira.geometry.tools import make_circle, signed_distance_2D_polygon, _signed_distance_2D
import numpy as np
import matplotlib.pyplot as plt
small_circle = make_circle(5, center=(0, 0, 0), axis=(0, 1, 0))
big_circle = make_circle(10, center=(0, 0, 0), axis=(0, 1, 0))

n = 100
x = z= np.linspace(-10, 10, n)
xx, zz = np.meshgrid(x, z, indexing="ij")
small_coords = small_circle.discretize(n).xz.T
all_coords = np.array([xx.flatten(), zz.flatten()]).T
sdfp = signed_distance_2D_polygon(all_coords, small_coords)
sdfp = sdfp.reshape((n, n))
f, ax = plt.subplots()
cm = ax.contourf(xx, zz, sdfp, cmap="RdBu")
ax.plot(*small_coords.T, color="r")
f.colorbar(cm)
ax.set_title("Signed distance function grid->circle")
ax.set_xlabel("x")
ax.set_ylabel("z")
ax.set_aspect("equal")
plt.show()
```
![image](https://github.com/Fusion-Power-Plant-Framework/bluemira/assets/26097289/b5f5822c-7d27-4d7d-8e6c-2f99b6211d82)

So any points inside the small coords (target poly) is positive, and any point outside is negative. Looking again at the docstring, and this is not what is written (!). This seems to be the main (and only?) problem and should be fixed. I will put in a quick documentation PR if you're happy with this explanation.

Basically the usage in a KOZ constraint should be: `signed_distance_2D_polygon(my_shape_coords, koz_coords) <= 0`

As a side note, I was initially extremely concerned that the values of your big-> small and small->big circles were not the same. Then I realised it is an artefact of the discretisation and the different radii; if you increase n, they will tend to be equal.

**Comment by hsaunders1904** [comment]
I think where my confusion was coming from, is what I assumed 'keep-out zone' meant. In the second case I mentioned above (big circle is the target, small circle is the keep-out zone), I was expecting that, because the keep-out zone is within the shape, it should be a violation of the constraint. But the signed distance comes out negative. 

Is the keep-out zone "the target does not intersect the zone"? And not "the union of the target and the zone's areas are 0"?

**Comment by CoronelBuendia** [comment]
> I think where my confusion was coming from, is what I assumed 'keep-out zone' meant. In the second case I mentioned above (big circle is the target, small circle is the keep-out zone), I was expecting that, because the keep-out zone is within the shape, it should be a violation of the constraint. But the signed distance comes out negative.
> 
> Is the keep-out zone "the target does not intersect the zone"? And not "the union of the target and the zone's areas are 0"?

If the shape is outside the keep-out zone, (i.e. the KOZ is fully enclosed by the shape), then the SDF is negative and the constraint is satisfied.

If the shape were partially or fully _inside_ the keep-out zone, this would return some or all positive values, and the constraint would be violated. Remember we're just talking about polygons (wires), not faces, so the concept of area does not exist here.

Sorry, I'm very tired. Does that make more sense?
